FHEM/00_SIGNALduino.pm

```diff
@@ -2528,7 +2528,7 @@ sub SIGNALduino_Parse_MU {
   my $hash = shift // return;    #return if no hash  is provided
   my $rmsg = shift // return;    #return if no rmsg is provided
   
-  if ($rmsg !~ /^MU;(?:P[0-7]=-?[0-9]+;){3,8}?((?:D=[0-7]+;)|(?:CP=[0-7];)|(?:R=\d+;)|(?:O;)?|(?:[A-Za-z0-9=]+;)*)*$/){   
+  if ($rmsg !~ /^(?=.*D=\d+)(MU;(?:P[0-7]=-?[0-9]{1,5};){2,8}((D=\d{2,};)|(CP=\d;)|(R=\d+;)?|(O;)?|(e;)?|(p;)?|(w=\d;)?)*)$/){
```

Comment:
Was hat es mit dem Teil auf sich? Die Werte sind mir nicht bekannt.

|(e;)?|(p;)?|(w=\d;)?

Reply:
Das sind Werte aus / von alter Firmware oder von der FW von Ralf. 
Diese Prüfung muss mit akzeptiert werden, da sonst dein Test `08_DeviceData_rmsg.t` fehl schlägt an Bsp. der Stelle

https://github.com/RFD-FHEM/SIGNALduino_TOOL/blob/3c6898029b9dbcef68dd76d7058ec241101e2dc5/FHEM/lib/SD_Device_ProtocolList.json#L434

oder

https://github.com/RFD-FHEM/SIGNALduino_TOOL/blob/3c6898029b9dbcef68dd76d7058ec241101e2dc5/FHEM/lib/SD_Device_ProtocolList.json#L1959

.... und andere Stellen.

Die Nachrichten im JSON zu ändern wäre der falsche Angriffspunkt.

Deine jetzige Prüfung https://github.com/RFD-FHEM/RFFHEM/blob/8aeae92e03b183c9d358c9a2aa614f205ba02597/FHEM/00_SIGNALduino.pm#L2531 lies diese Nachrichten auch durch.

Reply:
Also ich behaupte von der SIGNALduino Firmware kommen diese Daten nicht. 
Wieso sollten wir solche Daten im Modul verarbeiten. Im Prinzip kennen wir den Aufbau hier nicht.

Reply:
Es sind Daten welche wir im Laufe der Zeit akzeptiert haben wo zum damaligen Zeitpunkt mehrere Entwickler ein Projekt beisteuerten. Wenn du die Daten im JSON nun herausnimmst, sind 
- keine Testdaten mehr zur Verfügung
- google findet keine Ansätze mehr, wenn wir diese so eingeben aufgrund von Recherchen
- sämtliche Daten welche bisher akzeptiert wurden, auf einmal nicht mehr akptiert werden

Du in "deiner" bisherigen Prüfung diese auch akzeptierst in den Tests und auf einmal bei einer Verbesserung dies in Frage stellst, obwohl selbst dieses du schonmal abgenickt hattest ;-)

Reply:
Da hast Du Recht, die bisherige Regex hat alles von A-Z erlaubt. 

Die capture Groups habe ich mal entfernt, die brauchen wir nicht oder?

`^(?=.*D=\d+)(?:MU;(?:P[0-7]=-?[0-9]{1,5};){2,8}((?:D=\d{2,};)|(?:CP=\d;)|(?:R=\d+;)?|(?:O;)?|(?:e;)?|(?:p;)?|(?:w=\d;)?)*)$`

Was hat es mit dem positiven lookahead am Anfang auf sich, wozu dient der?

Reply:
Dein vorgeschlagenes Regex

Reply:
1)
Dein vorgeschlagenes Regex

`^(?=.*D=\d+)(?:MU;(?:P[0-7]=-?[0-9]{1,5};){2,8}((?:D=\d{2,};)|(?:CP=\d;)|(?:R=\d+;)?|(?:O;)?|(?:e;)?|(?:p;)?|(?:w=\d;)?)*)$`

sollte ebenso funktionieren. Ich habe es mal durchgetestet mit den markanten Messages.

Im Prinzip ist es ja das gleich. Ich habe sie mir mal gegenüber gestellt vereinfacht.
```
^(?=.*D=\d+)(  MU;(?:P[0-7]=-?[0-9]{1,5};){2,8}((  D=\d{2,};)|(  CP=\d;)|(  R=\d+;)?|(  O;)?|(  e;)?|(  p;)?|(  w=\d;)?)*)$
^(?=.*D=\d+)(?:MU;(?:P[0-7]=-?[0-9]{1,5};){2,8}((?:D=\d{2,};)|(?:CP=\d;)|(?:R=\d+;)?|(?:O;)?|(?:e;)?|(?:p;)?|(?:w=\d;)?)*)$
```

2)
`Was hat es mit dem positiven lookahead am Anfang auf sich, wozu dient der?`

Es dient dazu, das ein Datenteil vorhanden sein muss. Wenn wir den positiven lookahead nicht angeben, so wäre eine Nachricht wie
`MU;P0=11130;P1=-666;P2=533;P3=-273;P7=-6200;`
ein Match, was falsch ist.

ohne positiven lookahead am Anfang -> Siehe https://regex101.com/r/cWvI5R/1
mit positiven lookahead am Anfang -> Siehe https://regex101.com/r/1nRvRD/1

Grund dafür ist, das der Teil hinter den Pattern als ODER Gruppe angegeben ist weil die Reihenfolge variiert. (Daten im JSON)
Bsp: 
```
MU;P0=130;P1=-666;P2=533;P3=-273;P5=-6200;CP=0;R=15;D=01232301230123010101010101010123230501232323232323012323012301230101010101010101232305012323232323230123230123012301010101010101012323050123232323232301232301230123010101010101010123230501232323232323012323012301230101010101010101232305012323232323230123;O;
MU;P0=-392;P1=1285;P2=-1024;P3=506;P4=-27790;D=0121212123212321232323212323212123212321212321232123232121212121212321232123232121232121412121212321232123232321232321212321232121232123212323212121212121232123212323212123212141212121232123212323232123232121232123212123212321232321212121212123212321232;CP=3;R=0;O;
```

Sehr schön siehst du hier dir Aufklärung https://www.debuggex.com/r/16IIMljXVJO1V-Z-

Reply:
Danke für die sehr ausführliche und anschauliche Erklärung. Leuchtet alles ein, das mit dem oder hatte ich nicht im Blick. Super dass Du es gefunden hast.

Reply:
Bezüglich der capture groups, die Anpassung von mir verwendet halt non capturing groups, weil wir die capture groups nicht brauchen. Mehr Änderung war da nicht. Die Regex selbst habe ich nicht verändert.

Reply:
Da haben wir ja das RegEx "durchgekaut" :-)
Nimmst du die Änderung direkt in den Branch vor oder soll ich dies anpassen?

Reply:
Wäre mir Recht, wenn Du es anpasst

Reply:
Ich habe es angepasst